### ISSUE TITLE
"Slot" a field property, not a fieldOption

### ISSUE THREAD

**Issue by jettway** [jarvelov/vue-form-json-schema#9]
Docs claim that 'slot' is a fieldOption. I tried that to render an errorHandler with no success. I guessed that 'slot' might be a property of the field object instead, and that worked. 

vue-form-json-schema/docs/api/ui-schema/field/field-options/slot.md

**Comment by jarvelov** [comment]
That is really weird, I'm looking into this. Thanks for reporting this discrepancy!

**Comment by janschoepke** [comment]
Are there any news on this? Currently it is not possible for me to assign children to a given slot. In my case, it works neither through the fieldOptions object (as specified in the documentation) nor through the Field object itself, as suggested in this issue.

**Comment by janschoepke** [comment]
I just tried to find the place in the code where you read and process the slots. I searched for "slot" in the whole project and could find the following results: 

```./src/vfjs-field-component/index.js:      this.$slots.default,
./docs/api/ui-schema/field/error-handler.md:The component will get the the same props as any other component but it cannot receive any [children](children.md) as the default slot is populated by the rendered `errors`.
./docs/api/ui-schema/field/error-handler.md:The errors are passed as the default slot are rendered individually as a `<div>`.
./docs/api/ui-schema/field/field-options/slot.md:# slot
./docs/api/ui-schema/field/field-options/slot.md:> Used on content inserted into child components to indicate which named slot the content belongs to. - [Vue.js guide](https://vuejs.org/v2/api/#slot)
./docs/api/ui-schema/field/field-options/slot.md:  <slot name="main">
./docs/api/ui-schema/field/field-options/slot.md:    <span>This is the default main slot text</span>
./docs/api/ui-schema/field/field-options/slot.md:  </slot>
./docs/api/ui-schema/field/field-options/slot.md:              innerHTML: 'This child component will occupy the main slot of the parent'
./docs/api/ui-schema/field/field-options/slot.md:            slot: 'main'
./docs/api/ui-schema/field/field-options.md:* [slot](field-options/slot.md)
./docs/api/ui-schema/field/children.md:An array of [fields](../field.md). The rendered children (vnodes) will be passed in the default slot to the `component`.
./docs/SUMMARY.md:          * [slot](api/ui-schema/field/field-options/slot.md)
```

Except for the first result, which writes the default slot into the component by default, I did not find any results that are not part of the documentation. Where was this feature implemented? I couldn't find any reference to an existing implementation even in the commit messages.

**Comment by jarvelov** [comment]
Hello

In previous versions of Vue (< 2.6) the slot system worked a lot differently from how it does now and I haven't been able to get slots to work again since then, I'm sorry that this has not been reflected in the documentation.

The only way I've found to make slots kinda work is by using a `scopedSlot`, but this requires using JavaScript and does not generate any markup, but rather just returns a string:

```js
  component: {
    template: `<div>slot name="header" /></div>`
  },
  fieldOptions: {
    scopedSlots: {
      header: () => 'hello'
    }
  }
```

I want to bring back this functionality, however unfortunately I haven't had the time to dig into how to make it work again. I'll update the documentation tonight to say that slots are not currently working.

Oh, and since you wondered how slots even could work in the first place even though it is not mentioned: It's because all fieldOptions are passed into the render function, see for example [here](https://github.com/jarvelov/vue-form-json-schema/blob/master/src/vfjs-field-mixin/methods/getters.js#L59) how the rest of fieldOptions is assigned to the options variable and then returned as a top level property.

This means that any property in the fieldOptions will be passed to the render function.

**Comment by jarvelov** [comment]
I finally figured out how to fix the slots, so now it actually works they way it says in the documentation.  The fix is relased in `v2.5.2`. Please let me know if you're still having problems.

**Comment by janschoepke** [comment]
Awesome! I will test it tomorrow morning! Thanks in advance! :)

**Comment by janschoepke** [comment]
Hi @jarvelov, 
First of all, thanks again for the quick help! I have just tried the correction, unfortunately it does not seem to work as it should. I have updated the library to version 2.5.2, unfortunately the "slot" attribute is still not processed correctly, although I set the slot explicitly in the field options, vue-form-json-schema uses the "default" slot. Is there anything else I need to consider?
Here is an example, the component q-select occupies a slot named "after", which I want to set in the child object. This does not work as it should. 

`{
                        component: "q-select",
                        model: "testmodel",
                        fieldOptions: {
                            attrs: {
                                label: "Country",
                                options: [
                                    {
                                        "label": "Key 1",
                                        "value": "Value 1",
                                        "disable": false
                                    },
                                    {
                                        "label": "Key 2",
                                        "value": "Value 2",
                                        "disable": false
                                    }
                                ]
                            },
                            on: ["input"]
                        },
                        children: [
                            {
                                component: "span",
                                model: "testmodel2",
                                fieldOptions: {
                                    domProps: {
                                        innerHTML: "Test"
                                    },
                                    slot: "after"
                                }
                            }
                        ]
                    },`

**Comment by jarvelov** [comment]
I think it would be beneficial to see the source code of the `q-select` component. Is the `q-select` component a part of an open source library? If not or if it is your custom component can you please post the `<template>` of the `q-select` component.

I'll post back with an example of what worked for me when I head of to lunch in an hour or so.

**Comment by janschoepke** [comment]
q-select is the Select Field of the Quasar framework, which can be found here: 

https://quasar.dev/vue-components/select

**Comment by jarvelov** [comment]
I've created an example and replicated the problem. I might have found a temporary workaround for you.

Example: https://codesandbox.io/s/issue-9-h99p1

The problem occurs only when a `model` is used on the component we want to use in the slot. I'll have to dig a bit deeper but I think I could solve this pretty quickly.

However in your case there might be a workaround as I mentioned above.

You've add the `model` property on a `span`, which makes little sense as only components which emits events should have a `model` property on them. If you just want to show some text in the `after` slot you can remove the `model` property on the child to get it working. See the codesandbox link above.

The issue still needs to be fixed but hopefully you can use the workaround for now until I've fixed the issue completely.

**Comment by janschoepke** [comment]
@jarvelov 
Wow, your commitment is fantastic, thank you very much for this quick problem solving. As you have correctly recognized, it is pointless to insert a span into the slot. This only served as an example to make the testing and reproducing as easy as possible. In my case another component "QBtn" should be inserted into this slot. I have also recorded my scenario in a CodeSandbox: 

https://codesandbox.io/s/issue-9-pgg1d

As you can see the QBtn component is displayed incorrectly (in the default slot of the QSelect component). If I remove the model (commenting on line 92), the correct slot ("after") is occupied, but this seems to have an effect on the QBtn component, because the icon is no longer displayed. This means that the model is indispensable for correct functioning.

**Comment by jarvelov** [comment]
Well this is very interesting. I have not encountered an issue like one this before and this behavior is quite puzzling. Unfortunately I have to go back to work now but I hope I can dig into this this after work.

**Comment by alexravenna** [comment]
It looks like I've been trying to do something similar off and on for the past few weeks...nice to see someone else bring this up here 😄.

I'm trying to use a wrapper component (with slots) in the UI schema, where the actual component that I get input from (through `input`/`change`) is a child/slot in the wrapper component. When I try to attach the child/slot component to the VFJS model, then the component disappears and is replaced by the default slot content. Here's the CodeSandbox:

https://codesandbox.io/s/vfjs-slots-bdpd1

Line 57 in `MyEditor.vue` is where it's attempted to attach the child/slot component to the VFJS model.

**Comment by jarvelov** [comment]
I just published `v2.5.3` which should resolve the issues you are seeing @alexravenna 

Here's the working version of your code: https://codesandbox.io/s/vfjs-slots-s2goe
I have not made any changes other than upgrading to `v2.5.3`.

@janschoepke I am still investigating the issues you are having.

**Comment by jarvelov** [comment]
@janschoepke I figured out what the issue was. Here's a working version of your example code: https://codesandbox.io/s/issue-9-frm5f

With `v2.5.3`, which I just released, all that was needed to fix the missing icon in `q-btn` was to change `attrs` to `props`. With Vue the line between what is a component property (`prop`) and what is an attribute (`attr`) is blurred, but it makes a great difference when using the [render function](https://vuejs.org/v2/guide/render-function.html) which VFJS is using underneath the hood.

Most likely you always want to use `props` when it comes to Vue components.

`attrs` is for native HTML elements, like `input` where `type="tel"` is an example of an attribute. 

Then there's `domProps`, and well, I'm not entirely sure why it's a separate thing but it's where you set `innerHTML`. Vue has some [more info on this in their guide](https://vuejs.org/v2/guide/render-function.html#The-Data-Object-In-Depth) if anyone wants to dig deeper into where to put what property.
